snuba/query/snql/__init__.py

```diff
@@ -1,28 +1,194 @@
+    visit_function_call,
+    visit_function_name,
+    visit_high_pri_arithmetic,
+    visit_high_pri_op,
+    visit_high_pri_tuple,
+    visit_low_pri_arithmetic,
+    visit_low_pri_op,
+    visit_low_pri_tuple,
+    visit_numeric_literal,
+    visit_parameter,
+    visit_parameters_list,
+    visit_quoted_literal,
+)
+
+snql_grammar = Grammar(
+    fr"""
+    query_exp             = match_clause? where_clause? collect_clause? having_clause? order_by_clause?
+    match_clause          = space* "MATCH" space* "(" clause ")" space*
+    where_clause          = space* "WHERE" clause space*
+    collect_clause        = space* "COLLECT" collect_list space*
+    order_by_clause       = space* "ORDER BY" collect_list ("ASC"/"DESC") space*
 
+    collect_list          = collect_columns* (low_pri_arithmetic)
+    collect_columns       = low_pri_arithmetic space* comma space*
 
-grammar = Grammar(
-    r"""
-    query_exp = match_clause where_clause? collect_clause? having_clause? order_by_clause?
-    match_clause = space* "MATCH" "(" clause ")" space*
-    where_clause = space* "WHERE" clause space*
-    collect_clause = space* "COLLECT" clause "BY" clause space*
-    having_clause = space* "HAVING" clause space*
-    order_by_clause = space* "ORDER BY" clause ("ASC"/"DESC") space*
+    having_clause         = space* "HAVING" clause space*
+    clause                = space* ~r"[-=><\w]+" space*
+
+    low_pri_arithmetic    = space* high_pri_arithmetic space* (low_pri_tuple)*
+    high_pri_arithmetic   = space* arithmetic_term space* (high_pri_tuple)*
+    low_pri_tuple         = low_pri_op high_pri_arithmetic
+    high_pri_tuple        = high_pri_op arithmetic_term
+    arithmetic_term       = (space*) (function_call / numeric_literal / column_name) (space*)
+    low_pri_op            = "+" / "-"
+    high_pri_op           = "/" / "*"
+    param_expression      = low_pri_arithmetic / quoted_literal
+    parameters_list       = parameter* (param_expression)
+    parameter             = param_expression space* comma space*
+    function_call         = function_name open_paren parameters_list? close_paren (open_paren parameters_list? close_paren)?
+    simple_term           = quoted_literal / numeric_literal / column_name
+    literal               = ~r"[a-zA-Z0-9_\.:-]+"
+    quoted_literal        = "'" string_literal "'"
+    string_literal        = ~r"[a-zA-Z0-9_\.\+\*\/:-]*"
+    numeric_literal       = ~r"-?[0-9]+(\.[0-9]+)?(e[\+\-][0-9]+)?"
+    column_name           = ~r"[a-zA-Z_][a-zA-Z0-9_\.]*"
+    function_name         = ~r"[a-zA-Z_][a-zA-Z0-9_]*"
+    open_paren            = "("
+    close_paren           = ")"
+    space                 = " "
+    comma                 = ","
 
-    clause = space* ~r"[-=><\w]+" space*
-    space = " "
 """
 )
 
 
+class SnQLVisitor(NodeVisitor):
+    """
+    Builds Snuba AST expressions from the Parsimonious parse tree.
+    """
+
+    def visit_query_exp(self, node, visited_children):
```

Comment:
Please add the missing type signatures to all methods.

Reply:
I will get to this asap! Just wanted to update the GROUP BY and ORDER BY additions first.

Reply:
still types missing.